File: swift/Tests/HyperCastTests/CastTests.swift
import Foundation
import XCTest
@testable import HyperCast

/// Binding-level behavior the corpus can't express: the enum union's mandatory-exhaustive
/// switch, locale bridging, Swift-type fidelity (native unsigned, attosecond-backed
/// `Duration`, digit-perfect `DateComponents`), and the caller-bug guards.
final class CastTests: XCTestCase {
    /// The core crate's own `version = "..."` from `rust/Cargo.toml`, found by walking up
    /// from this file the way `CorpusTests` finds `corpus/` — so the version assertion
    /// follows a release bump instead of going stale on it.
    private static let crateVersion: String = {
        var dir = URL(fileURLWithPath: #filePath)
        while true {
            let manifest = dir.appendingPathComponent("rust/Cargo.toml")
            if let text = try? String(contentsOf: manifest, encoding: .utf8) {
                // Split on any newline, not on "\n": Swift treats "\r\n" as one Character, so a
                // CRLF checkout (Windows) would otherwise read as a single line. Bit CI for real.
                for line in text.split(whereSeparator: \.isNewline) where line.hasPrefix("version = \"") {
                    return String(line.dropFirst("version = \"".count).prefix { $0 != "\"" })
                }
                fatalError("no version = \"...\" line in \(manifest.path)")
            }
            let parent = dir.deletingLastPathComponent()
            if parent.path == dir.path { break }
            dir = parent
        }
        fatalError("rust/Cargo.toml not found above \(#filePath)")
    }()

    func testVerdictSwitchIsExhaustiveWithTwoCases() throws {
        // This compiling at all is the feature — and in Swift it's not opt-in: the
        // compiler rejects a non-exhaustive switch over an enum, full stop.
        let rendered: String = switch try Cast.i32("42", format: .invariant) {
        case .success(let value): "ok \(value)"
        case .fault(let fault): "fault \(fault.reason)"
        }
        XCTAssertEqual(rendered, "ok 42")
    }

    func testFaultSpanPointsAtTheOffendingByte() throws {
        XCTAssertEqual(
            try Cast.i32("  12x4", format: .invariant),
            .fault(Fault(reason: .malformed, offset: 4, length: 1)))
    }

    func testStringDoorsTranscodeNonAsciiSeparators() throws {
        let french = NumFormat(decimalSeparator: ",", groupSeparator: " ", styles: .all)
        XCTAssertEqual(try Cast.f64("1 234,5", format: french), .success(1234.5))
    }

    func testLocaleBridge() throws {
        let german = NumFormat.from(locale: Locale(identifier: "de_DE"))
        XCTAssertEqual(german.decimalSeparator, ",")
        XCTAssertEqual(german.groupSeparator, ".")
        XCTAssertEqual(try Cast.f64("1.234,5", format: german), .success(1234.5))
    }

    func testCurrencySymbolComesFromTheLocale() throws {
        // en_US's own symbol, at every edge the core admits: leading, after the sign,
        // before it with whitespace, and wrapped by accounting parentheses.
        let enUs = NumFormat.from(locale: Locale(identifier: "en_US"))
        XCTAssertEqual(enUs.currencySymbol, "$")
        XCTAssertTrue(enUs.styles.contains(.currency))
        XCTAssertEqual(try Cast.i32("$1,234", format: enUs), .success(1234))
        XCTAssertEqual(try Cast.i32("-$5", format: enUs), .success(-5))
        XCTAssertEqual(try Cast.f64("$ -5", format: enUs), .success(-5))
        XCTAssertEqual(try Cast.i32("($5)", format: enUs), .success(-5))
        XCTAssertEqual(try Cast.decimal("($1,234.50)", format: enUs), .success(Decimal(string: "-1234.50")!))
    }

    func testTrailingMultibyteCurrencySymbol() throws {
        // A declared symbol is matched whole at the edge, so one that happens to contain
        // the group separator (`kr.` under `.` grouping) is no trouble.
        let danish = NumFormat(decimalSeparator: ",", groupSeparator: ".", styles: .all, currencySymbol: "kr.")
        XCTAssertEqual(try Cast.decimal("1.234,50 kr.", format: danish), .success(Decimal(string: "1234.50")!))
        let euro = NumFormat(decimalSeparator: ",", groupSeparator: ".", styles: .all, currencySymbol: "€")
        XCTAssertEqual(try Cast.i32("5 €", format: euro), .success(5))
    }

    func testCurrencySymbolFillsTheNativeFieldEndToEnd() throws {
        // Sixteen bytes exactly — both words of the packed field have to reach the core
        // intact for the symbol to be recognized at all.
        let widest = String(repeating: "¤", count: 8)
        XCTAssertEqual(widest.utf8.count, 16)
        let format = NumFormat(decimalSeparator: ".", groupSeparator: ",", styles: .all, currencySymbol: widest)
        XCTAssertEqual(try Cast.i32("\(widest)42", format: format), .success(42))
        XCTAssertEqual(try Cast.i32("42 \(widest)", format: format), .success(42))
    }

    func testDeclaredSymbolWithTheFlagOffIsMalformedAtTheSymbol() throws {
        let declaredOnly = NumFormat(
            decimalSeparator: ".", groupSeparator: ",", styles: NumStyles.all.subtracting(.currency),
            currencySymbol: "$")
        XCTAssertEqual(
            try Cast.i32("$5", format: declaredOnly),
            .fault(Fault(reason: .malformed, offset: 0, length: 1)))
        // And the flag with nothing declared is a no-op — the invariant profile.
        XCTAssertEqual(NumFormat.invariant.currencySymbol, "")
        XCTAssertEqual(
            try Cast.i32("$5", format: .invariant),
            .fault(Fault(reason: .malformed, offset: 0, length: 1)))
    }

    func testDecimalIsExactWhereADoubleIsNot() throws {
        XCTAssertEqual(try Cast.decimal("0.1", format: .invariant), .success(Decimal(string: "0.1")!))
        XCTAssertEqual(try Cast.decimal("-0.025", format: .invariant), .success(Decimal(string: "-0.025")!))
        XCTAssertEqual(try Cast.decimal("50%", format: .invariant), .success(Decimal(string: "0.5")!))
        // The full 96 bits, above and below the high word's 2⁶⁴ boundary, and the widest scale.
        XCTAssertEqual(
            try Cast.decimal("79228162514264337593543950335", format: .invariant),
            .success(Decimal(string: "79228162514264337593543950335")!))
        XCTAssertEqual(
            try Cast.decimal("-7.9228162514264337593543950335", format: .invariant),
            .success(Decimal(string: "-7.9228162514264337593543950335")!))
        XCTAssertEqual(
            try Cast.decimal("18446744073709551616", format: .invariant),
            .success(Decimal(string: "18446744073709551616")!))
        // Never rounds: a 29th place is out of range, not approximated.
        XCTAssertEqual(
            try Cast.decimal("0.00000000000000000000000000001", format: .invariant),
            .fault(Fault(reason: .outOfRange, offset: 0, length: 31)))
    }

    func testDecimalIsCanonical() throws {
        // The core trims exact trailing fraction zeros, so `1.10`, `1.1` and `1.1000` are
        // all magnitude 11 at scale 1 — exactly the form Foundation's Decimal represents.
        let canonical = Decimal(sign: .plus, exponent: -1, significand: 11)
        for text in ["1.10", "1.1", "1.1000"] {
            XCTAssertEqual(try Cast.decimal(text, format: .invariant), .success(canonical), text)
        }
        XCTAssertEqual(
            try Cast.decimal("1.0000000000000000000000000000000", format: .invariant),
            .success(Decimal(1)))
    }

    func testDecimalZeroIsNeverNegative() throws {
        guard case .success(let value) = try Cast.decimal("-0.00", format: .invariant) else {
            return XCTFail("-0.00 should parse")
        }
        XCTAssertTrue(value.isZero)
        XCTAssertEqual(value.sign, .plus)
        XCTAssertEqual(value, Decimal(string: "0.00")!)
    }

    func testNativeVersionIsTheLoadedLibrarysOwn() throws {
        XCTAssertEqual(try Cast.nativeVersion(), Self.crateVersion)
    }

    func testIsAvailableAgreesWithTheLoad() throws {
        XCTAssertTrue(Cast.isAvailable)
        XCTAssertNoThrow(try Cast.nativeVersion())
    }

    func testGenericNumericDoorReachesEveryTarget() throws {
        let invariant = NumFormat.invariant
        XCTAssertEqual(try Cast.numeric("-8", format: invariant), Verdict<Int8>.success(-8))
        XCTAssertEqual(try Cast.numeric("-16", format: invariant), Verdict<Int16>.success(-16))
        XCTAssertEqual(try Cast.numeric("(32)", format: invariant), Verdict<Int32>.success(-32))
        XCTAssertEqual(try Cast.numeric("-64", format: invariant), Verdict<Int64>.success(-64))
        XCTAssertEqual(try Cast.numeric("255", format: invariant), Verdict<UInt8>.success(255))
        XCTAssertEqual(try Cast.numeric("65535", format: invariant), Verdict<UInt16>.success(65535))
        XCTAssertEqual(try Cast.numeric("0xFFFFFFFF", format: invariant), Verdict<UInt32>.success(.max))
        XCTAssertEqual(
            try Cast.numeric("18446744073709551615", format: invariant), Verdict<UInt64>.success(.max))
        XCTAssertEqual(try Cast.numeric("1.5", format: invariant), Verdict<Float>.success(1.5))
        XCTAssertEqual(try Cast.numeric("50%", format: invariant), Verdict<Double>.success(0.5))
        XCTAssertEqual(
            try Cast.numeric("0.1", format: invariant), Verdict<Decimal>.success(Decimal(string: "0.1")!))
        // Each target keeps its own door's rules: a decimal point is never an integer, and
        // the integer door faults at the point itself.
        XCTAssertEqual(
            try Cast.numeric("1.5", format: invariant),
            Verdict<Int32>.fault(Fault(reason: .malformed, offset: 1, length: 1)))
        // The bytes and buffer forms route the same way.
        XCTAssertEqual(try Cast.numeric(Array("42".utf8), format: invariant), Verdict<UInt8>.success(42))
    }

    func testGenericNumericDoorPassesFaultsThrough() throws {
        let verdict: Verdict<Int32> = try Cast.numeric("  12x4", format: .invariant)
        XCTAssertEqual(verdict, .fault(Fault(reason: .malformed, offset: 4, length: 1)))
        let empty: Verdict<Decimal> = try Cast.numeric("   ", format: .invariant)
        XCTAssertNil(Cast.optional(empty))
    }

    func testRawNumFormatIsTheNativeStructsThirtyTwoBytes() {
        // u32 × 4 at 0/4/8/12, then the 16 symbol bytes at 16: 32 bytes, no padding
        // between the words. Swift's 8-byte alignment for the tuple is stricter than the
        // native struct's 4, which is compatible in the caller-allocated direction.
        XCTAssertEqual(MemoryLayout<Cast.RawNumFormat>.size, 32)
        XCTAssertEqual(MemoryLayout<Cast.RawNumFormat>.stride, 32)
    }

    func testCurrencySymbolPacksAsItsUtf8Bytes() {
        // `kr.` is 0x6B 0x72 0x2E; the low word's in-memory bytes must be exactly those.
        let danish = NumFormat(decimalSeparator: ",", groupSeparator: ".", styles: .all, currencySymbol: "kr.")
        XCTAssertEqual(danish.currencyLength, 3)
        XCTAssertEqual(withUnsafeBytes(of: danish.currencyLow) { Array($0.prefix(3)) }, Array("kr.".utf8))
        XCTAssertEqual(danish.currencyHigh, 0)
        XCTAssertEqual(NumFormat.invariant.currencyLength, 0)
    }

    func testUuidAgreesWithThePlatformsOwnParser() throws {
        let text = "01020304-0506-0708-090a-0b0c0d0e0f10"
        XCTAssertEqual(try Cast.uuid(text), .success(UUID(uuidString: text)!))
        XCTAssertEqual(try Cast.uuid("urn:uuid:\(text)"), .success(UUID(uuidString: text)!))
    }

    func testUnsignedDoorsAreNativelyUnsigned() throws {
        XCTAssertEqual(try Cast.u8("255", format: .invariant), .success(UInt8(255)))
        XCTAssertEqual(try Cast.u64("18446744073709551615", format: .invariant), .success(UInt64.max))
    }

    func testTimestampNormalizesToUtc() throws {
        XCTAssertEqual(
            try Cast.timestamp("2026-01-02T15:04:05+05:00"),
            .success(Date(timeIntervalSince1970: 1_767_348_245)))
    }

    func testUnixMapsTheDeclaredPrecision() throws {
        XCTAssertEqual(
            try Cast.unix("-1", precision: .seconds),
            .success(Date(timeIntervalSince1970: -1)))
        XCTAssertEqual(
            try Cast.unix("1700000000123", precision: .milliseconds),
            .success(Date(timeIntervalSince1970: 1_700_000_000.123)))
    }

    func testDateTimeAndDurationKeepFullFidelity() throws {
        XCTAssertEqual(
            try Cast.date("2026-01-02"),
            .success(DateComponents(year: 2026, month: 1, day: 2)))
        XCTAssertEqual(
            try Cast.time("15:04:05.123456789"),
            .success(DateComponents(hour: 15, minute: 4, second: 5, nanosecond: 123_456_789)))
        // Duration is attosecond-backed — the core's nanoseconds carry exactly, both signs.
        XCTAssertEqual(try Cast.duration("P1DT6H"), .success(.seconds(108_000)))
        XCTAssertEqual(
            try Cast.duration("-1.5s"),
            .success(Duration.seconds(-1) + .nanoseconds(-500_000_000)))
        XCTAssertEqual(
            try Cast.duration("00:00:00.123456789"),
            .success(Duration.nanoseconds(123_456_789)))
    }

    func testOptionalPresentsEmptyAsNil() throws {
        XCTAssertNil(Cast.optional(try Cast.i32("   ", format: .invariant)))
        XCTAssertEqual(Cast.optional(try Cast.i32("42", format: .invariant)), .success(42))
        XCTAssertEqual(
            Cast.optional(try Cast.i32("abc", format: .invariant)),
            .fault(Fault(reason: .malformed, offset: 0, length: 1)))
    }

    func testBytesAndStringDoorsAgree() throws {
        XCTAssertEqual(try Cast.bool("yes"), try Cast.bool(Array("yes".utf8)))
        XCTAssertEqual(try Cast.bool("yes"), .success(true))
    }
    func testDateOrderDisambiguatesLikeTheCulturesDo() throws {
        // The canonical ambiguity: 1/7/2026 is January 7th under en-US's month-first short
        // dates and July 1st under en-GB's day-first ones — resolved only by declaration,
        // with the declaration derived from the real locales' own short-date patterns.
        let enUs = DateOrder.from(locale: Locale(identifier: "en_US"))
        let enGb = DateOrder.from(locale: Locale(identifier: "en_GB"))
        XCTAssertEqual(enUs, .monthDayYear)
        XCTAssertEqual(enGb, .dayMonthYear)
        guard case .success(let us) = try Cast.date("1/7/2026", order: enUs) else {
            return XCTFail("en-US order should parse")
        }
        XCTAssertEqual(us, DateComponents(year: 2026, month: 1, day: 7))
        guard case .success(let gb) = try Cast.date("1/7/2026", order: enGb) else {
            return XCTFail("en-GB order should parse")
        }
        XCTAssertEqual(gb, DateComponents(year: 2026, month: 7, day: 1))
        // Undeclared, the door stays strict ISO — the ambiguity is never guessed at.
        guard case .fault(let fault) = try Cast.date("1/7/2026") else {
            return XCTFail("undeclared slash date should be malformed")
        }
        XCTAssertEqual(fault.reason, .malformed)
    }

    func testDateTimeReadsTheMessyCivilShapes() throws {
        // The AM/PM world, zone-less: no timeZone component because the text named no
        // zone — fusing one is the caller's job, never the parser's guess.
        guard case .success(let us) = try Cast.dateTime("1/7/2026 3:04 PM", order: .monthDayYear) else {
            return XCTFail("en-US order should parse")
        }
        XCTAssertEqual(
            us, DateComponents(year: 2026, month: 1, day: 7, hour: 15, minute: 4, second: 0, nanosecond: 0))
        XCTAssertNil(us.timeZone)
        guard case .success(let gb) = try Cast.dateTime("1/7/2026 3:04 PM", order: .dayMonthYear) else {
            return XCTFail("en-GB order should parse")
        }
        XCTAssertEqual(gb.month, 7)
        XCTAssertEqual(gb.day, 1)
        // A zone suffix is not this door's business — timestamp is the instant door.
        guard case .fault(let fault) = try Cast.dateTime("1/7/2026 15:04:05Z", order: .monthDayYear) else {
            return XCTFail("zoned text should be malformed through the civil door")
        }
        XCTAssertEqual(fault.reason, .malformed)
    }

}

File: csharp/HyperCast.Tests/CastTests.cs
using System.Globalization;

namespace HyperCast.Tests;

/// <summary>
/// Binding-level behavior the corpus can't express: the union's compile-checked consumption,
/// culture-to-<see cref="NumFormat"/> bridging, .NET-type fidelity mapping, the UTF-16 doors'
/// transcode path, and the caller-bug guards.
/// </summary>
public sealed class CastTests
{
	[Fact]
	void Verdict_switch_expression_is_exhaustive_with_two_arms()
	{
		// This compiling at all is the feature: CS8509 is an error in this repo, so these
		// two arms with no default are proven exhaustive by the compiler.
		var rendered = Cast.Int32("42", NumFormat.Invariant) switch
		{
			Success<int> success => $"ok {success.Value}",
			Fault fault => $"fault {fault.Reason}",
		};
		rendered.ShouldBe("ok 42");
	}

	[Fact]
	void Verdict_pattern_matches_case_types_directly()
	{
		(Cast.Boolean("yes") is Success<bool> { Value: true }).ShouldBeTrue();
		(Cast.Boolean("maybe") is Fault { Reason: CastFailure.Malformed }).ShouldBeTrue();
	}

	[Fact]
	void Defaulted_verdict_matches_neither_case()
	{
		var defaulted = default(Verdict<int>);
		defaulted.HasValue.ShouldBeFalse();
		(defaulted is Success<int>).ShouldBeFalse();
		(defaulted is Fault).ShouldBeFalse();
	}

	[Fact]
	void Defaulted_fault_is_rejected_as_a_case_value() =>
		Should.Throw<ArgumentOutOfRangeException>(() => new Verdict<int>(default(Fault)));

	[Fact]
	void NumFormat_bridges_from_a_real_culture()
	{
		var german = NumFormat.From(CultureInfo.GetCultureInfo("de-DE"));
		german.DecimalSeparator.ShouldBe(',');
		german.GroupSeparator.ShouldBe('.');
		(Cast.Double("1.234,5", german) is Success<double> { Value: 1234.5 }).ShouldBeTrue();
	}

	[Fact]
	void Utf16_doors_transcode_non_ascii_separators()
	{
		var french = new NumFormat(',', ' ', NumStyles.All);
		(Cast.Double("1 234,5", french) is Success<double> { Value: 1234.5 }).ShouldBeTrue();
	}

	[Fact]
	void Utf16_doors_survive_inputs_past_the_stack_buffer()
	{
		var oversized = new string('x', 4096);
		(Cast.Boolean(oversized) is Fault { Reason: CastFailure.Malformed, Offset: 0, Length: 4096 })
			.ShouldBeTrue();
	}

	[Fact]
	void Fault_span_points_at_the_offending_byte()
	{
		(Cast.Int32("  12x4", NumFormat.Invariant) is Fault { Reason: CastFailure.Malformed, Offset: 4, Length: 1 })
			.ShouldBeTrue();
	}

	[Fact]
	void Uuid_agrees_with_the_platforms_own_parser()
	{
		const string text = "01020304-0506-0708-090a-0b0c0d0e0f10";
		Cast.Uuid(text).ShouldBe(new Verdict<Guid>(new Success<Guid>(Guid.Parse(text))));
		Cast.Uuid($"urn:uuid:{text}").ShouldBe(new Verdict<Guid>(new Success<Guid>(Guid.Parse(text))));
	}

	[Fact]
	void Timestamp_maps_to_a_utc_datetimeoffset_at_tick_fidelity()
	{
		var expected = new DateTimeOffset(2026, 1, 2, 15, 4, 5, TimeSpan.Zero).AddTicks(1234567);
		(Cast.Timestamp("2026-01-02T15:04:05.1234567Z") is Success<DateTimeOffset> success).ShouldBeTrue();
		Cast.Timestamp("2026-01-02T15:04:05.1234567Z").ShouldBe(new(new Success<DateTimeOffset>(expected)));
		// Offset input normalizes to UTC — 15:04:05+05:00 is 10:04:05Z.
		Cast.Timestamp("2026-01-02T15:04:05+05:00")
			.ShouldBe(new(new Success<DateTimeOffset>(new(2026, 1, 2, 10, 4, 5, TimeSpan.Zero))));
		// Sub-tick nanoseconds (the core's extra fidelity) truncate to the tick.
		Cast.Timestamp("2026-01-02T15:04:05.123456789Z")
			.ShouldBe(new(new Success<DateTimeOffset>(
				new DateTimeOffset(2026, 1, 2, 15, 4, 5, TimeSpan.Zero).AddTicks(1234567))));
	}

	[Fact]
	void Unix_maps_the_declared_precision()
	{
		Cast.Unix("-1", UnixPrecision.Seconds)
			.ShouldBe(new(new Success<DateTimeOffset>(DateTimeOffset.UnixEpoch.AddSeconds(-1))));
		Cast.Unix("1700000000123", UnixPrecision.Milliseconds)
			.ShouldBe(new(new Success<DateTimeOffset>(DateTimeOffset.FromUnixTimeMilliseconds(1_700_000_000_123))));
	}

	[Fact]
	void Date_order_disambiguates_like_the_cultures_do()
	{
		// The canonical ambiguity: 1/7/2026 is January 7th under en-US's month-first short
		// dates and July 1st under en-GB's day-first ones — resolved only by declaration,
		// with the declaration derived from the real cultures' own short-date patterns.
		var enUs = DateOrders.From(CultureInfo.GetCultureInfo("en-US"));
		var enGb = DateOrders.From(CultureInfo.GetCultureInfo("en-GB"));
		enUs.ShouldBe(DateOrder.MonthDayYear);
		enGb.ShouldBe(DateOrder.DayMonthYear);
		(Cast.Date("1/7/2026", enUs) is Success<DateOnly> { Value: var us } && us == new DateOnly(2026, 1, 7))
			.ShouldBeTrue();
		(Cast.Date("1/7/2026", enGb) is Success<DateOnly> { Value: var gb } && gb == new DateOnly(2026, 7, 1))
			.ShouldBeTrue();
		// ISO-order cultures land on YearMonthDay; the ISO form is that order's subset.
		DateOrders.From(CultureInfo.GetCultureInfo("ja-JP")).ShouldBe(DateOrder.YearMonthDay);
		// Undeclared, the door stays strict ISO — the ambiguity is never guessed at.
		(Cast.Date("1/7/2026") is Fault { Reason: CastFailure.Malformed }).ShouldBeTrue();
		Should.Throw<ArgumentOutOfRangeException>(() => Cast.Date("1/7/2026", DateOrder.Unspecified));
	}

	[Fact]
	void Datetime_reads_the_messy_civil_shapes()
	{
		// The AM/PM world, zone-less: Kind is honestly Unspecified because the text named
		// no zone — fusing one is the caller's job, never the parser's guess.
		var enUs = DateOrders.From(CultureInfo.GetCultureInfo("en-US"));
		(Cast.DateTime("1/7/2026 3:04 PM", enUs) is Success<DateTime>
			{ Value: { Kind: DateTimeKind.Unspecified } value } && value == new DateTime(2026, 1, 7, 15, 4, 0))
			.ShouldBeTrue();
		(Cast.DateTime("1/7/2026 3:04 PM", DateOrder.DayMonthYear) is Success<DateTime> { Value: var gb }
			&& gb == new DateTime(2026, 7, 1, 15, 4, 0)).ShouldBeTrue();
		// A zone suffix is not this door's business — Timestamp is the instant door.
		(Cast.DateTime("1/7/2026 15:04:05Z", enUs) is Fault { Reason: CastFailure.Malformed }).ShouldBeTrue();
	}

	[Fact]
	void Date_time_and_duration_map_to_their_dotnet_types()
	{
		Cast.Date("2026-01-02").ShouldBe(new(new Success<DateOnly>(new(2026, 1, 2))));
		Cast.Time("15:04:05.1234567").ShouldBe(new(new Success<TimeOnly>(
			new TimeOnly(15, 4, 5).Add(TimeSpan.FromTicks(1234567)))));
		Cast.Duration("P1DT6H").ShouldBe(new(new Success<TimeSpan>(new(1, 6, 0, 0))));
		Cast.Duration("-1.5s").ShouldBe(new(new Success<TimeSpan>(TimeSpan.FromSeconds(-1.5))));
		Cast.Duration("01:30").ShouldBe(new(new Success<TimeSpan>(new(1, 30, 0))));
	}

	[Fact]
	void Optional_presents_empty_as_absent_and_everything_else_verbatim()
	{
		Cast.Optional(Cast.Int32("   ", NumFormat.Invariant)).ShouldBeNull();
		Cast.Optional(Cast.Int32("42", NumFormat.Invariant))
			.ShouldBe(new Verdict<int>(new Success<int>(42)));
		Cast.Optional(Cast.Int32("abc", NumFormat.Invariant))!.Value
			.TryGetValue(out Fault fault).ShouldBeTrue();
		fault.Reason.ShouldBe(CastFailure.Malformed);
	}

	[Fact]
	void NumFormat_bridges_from_any_format_provider()
	{
		// The shape every BCL TryParse takes — no CultureInfo cast at the call site.
		IFormatProvider provider = CultureInfo.GetCultureInfo("de-DE");
		NumFormat.From(provider).ShouldBe(NumFormat.From(CultureInfo.GetCultureInfo("de-DE")));
		NumFormat.From(NumberFormatInfo.InvariantInfo).CurrencySymbol.ShouldBe("¤");
		NumFormat.From(CultureInfo.InvariantCulture).ShouldBe(NumFormat.From(NumberFormatInfo.InvariantInfo));
	}

	[Fact]
	void Currency_symbol_comes_from_the_culture_and_is_matched_at_either_edge()
	{
		var usd = NumFormat.From(CultureInfo.GetCultureInfo("en-US"));
		usd.CurrencySymbol.ShouldBe("$");
		(Cast.Int32("$1,234", usd) is Success<int> { Value: 1234 }).ShouldBeTrue();
		(Cast.Int32("-$5", usd) is Success<int> { Value: -5 }).ShouldBeTrue();
		(Cast.Int32("($5)", usd) is Success<int> { Value: -5 }).ShouldBeTrue();
		(Cast.Double("$1,234.50", usd) is Success<double> { Value: 1234.5 }).ShouldBeTrue();
		(Cast.Decimal("$1,234.50", usd) is Success<decimal> { Value: 1234.50m }).ShouldBeTrue();
		var euro = NumFormat.From(CultureInfo.GetCultureInfo("de-DE"));
		euro.CurrencySymbol.ShouldBe("€");
		(Cast.Double("1.234,50 €", euro) is Success<double> { Value: 1234.5 }).ShouldBeTrue();
		// Declared but not permitted: the symbol is a stray character like any other.
		var declaredOff = usd with { Styles = NumStyles.All & ~NumStyles.Currency };
		(Cast.Int32("$5", declaredOff) is Fault { Reason: CastFailure.Malformed, Offset: 0, Length: 1 }).ShouldBeTrue();
		// Permitted but nothing declared: the invariant profile is unchanged by the flag.
		(Cast.Int32("$5", NumFormat.Invariant) is Fault { Reason: CastFailure.Malformed }).ShouldBeTrue();
	}

	[Fact]
	void Currency_symbol_that_would_collide_with_the_scan_is_a_caller_bug()
	{
		Should.Throw<ArgumentException>(() => Cast.Int32("5", NumFormat.Invariant with { CurrencySymbol = "US 1" }));
		Should.Throw<ArgumentException>(() => Cast.Int32("5", NumFormat.Invariant with { CurrencySymbol = new string('€', 6) }));
		// Multi-byte, multi-character symbols that fit are fine — and may contain a separator.
		var danish = new NumFormat(',', '.', NumStyles.All, "kr.");
		(Cast.Int64("1.234.567 kr.", danish) is Success<long> { Value: 1_234_567 }).ShouldBeTrue();
	}

	[Fact]
	void Decimal_is_exact_and_canonical()
	{
		Cast.Decimal("0.1", NumFormat.Invariant).ShouldBe(new(new Success<decimal>(0.1m)));
		// Canonical: trailing fraction zeros are trimmed, so the scale is minimal.
		Cast.Decimal("1.10", NumFormat.Invariant).TryGetValue(out Success<decimal> scaled).ShouldBeTrue();
		scaled.Value.ShouldBe(1.1m);
		scaled.Value.Scale.ShouldBe((byte)1);
		Cast.Decimal("100", NumFormat.Invariant).TryGetValue(out Success<decimal> hundred).ShouldBeTrue();
		hundred.Value.Scale.ShouldBe((byte)0);
		Cast.Decimal("(2.5)%", NumFormat.Invariant).ShouldBe(new(new Success<decimal>(-0.025m)));
		Cast.Decimal("79228162514264337593543950335", NumFormat.Invariant).ShouldBe(new(new Success<decimal>(decimal.MaxValue)));
		(Cast.Decimal("79228162514264337593543950336", NumFormat.Invariant) is Fault { Reason: CastFailure.OutOfRange }).ShouldBeTrue();
		(Cast.Decimal("0.00000000000000000000000000001", NumFormat.Invariant) is Fault { Reason: CastFailure.OutOfRange }).ShouldBeTrue();
		// Zero is never negative — no -0 decimal escapes the door.
		Cast.Decimal("-0.00", NumFormat.Invariant).TryGetValue(out Success<decimal> zero).ShouldBeTrue();
		decimal.IsNegative(zero.Value).ShouldBeFalse();
		zero.Value.Scale.ShouldBe((byte)0);
		(Cast.Decimal("1.234,50", NumFormat.Detect) is Success<decimal> { Value: 1234.50m }).ShouldBeTrue();
	}

	[Fact]
	void Numeric_dispatches_every_target_through_one_generic_door()
	{
		(Cast.Numeric<sbyte>("-128", NumFormat.Invariant) is Success<sbyte> { Value: -128 }).ShouldBeTrue();
		(Cast.Numeric<short>("(1,234)", NumFormat.Invariant) is Success<short> { Value: -1234 }).ShouldBeTrue();
		(Cast.Numeric<int>("0x2A", NumFormat.Invariant) is Success<int> { Value: 42 }).ShouldBeTrue();
		(Cast.Numeric<long>("1e3", NumFormat.Invariant) is Success<long> { Value: 1000 }).ShouldBeTrue();
		(Cast.Numeric<byte>("255", NumFormat.Invariant) is Success<byte> { Value: 255 }).ShouldBeTrue();
		(Cast.Numeric<ushort>("65535", NumFormat.Invariant) is Success<ushort> { Value: 65535 }).ShouldBeTrue();
		(Cast.Numeric<uint>("4294967295", NumFormat.Invariant) is Success<uint> { Value: 4294967295 }).ShouldBeTrue();
		(Cast.Numeric<ulong>("18446744073709551615", NumFormat.Invariant) is Success<ulong> { Value: ulong.MaxValue }).ShouldBeTrue();
		(Cast.Numeric<float>("2.5", NumFormat.Invariant) is Success<float> { Value: 2.5f }).ShouldBeTrue();
		(Cast.Numeric<double>("25.5%", NumFormat.Invariant) is Success<double> { Value: 0.255 }).ShouldBeTrue();
		(Cast.Numeric<decimal>("1.10", NumFormat.Invariant) is Success<decimal> { Value: 1.10m }).ShouldBeTrue();
		// The fault travels through untouched.
		(Cast.Numeric<byte>("256", NumFormat.Invariant) is Fault { Reason: CastFailure.OutOfRange }).ShouldBeTrue();
		Should.Throw<NotSupportedException>(() => Cast.Numeric<Int128>("1", NumFormat.Invariant));
	}

	[Fact]
	void Utf16_doors_report_char_offsets_not_byte_offsets()
	{
		// "€" is three UTF-8 bytes but one char: the byte door says 3+1, the char door 1+1.
		var euroThenX = "€x"u8;
		(Cast.Int32(euroThenX, NumFormat.Invariant) is Fault { Offset: 0, Length: 3 }).ShouldBeTrue();
		(Cast.Int32("€x", NumFormat.Invariant) is Fault { Offset: 0, Length: 1 }).ShouldBeTrue();
		(Cast.Int32("1€", NumFormat.Invariant) is Fault { Offset: 1, Length: 1 }).ShouldBeTrue();
		(Cast.Int32("1€"u8, NumFormat.Invariant) is Fault { Offset: 1, Length: 3 }).ShouldBeTrue();
		(Cast.Int32("€€1x", NumFormat.Invariant) is Fault { Offset: 0, Length: 1 }).ShouldBeTrue();
		(Cast.Double("1 234,5x", new NumFormat(',', ' ', NumStyles.All)) is Fault { Offset: 7, Length: 1 }).ShouldBeTrue();
	}

	[Fact]
	void Native_library_is_probed_once_and_reports_its_version()
	{
		Cast.IsAvailable.ShouldBeTrue();
		Cast.NativeVersion.ShouldNotBeNull();
		// The binding and the core share one version — the probe exists to catch the mismatch.
		var binding = typeof(Cast).Assembly.GetName().Version!;
		Cast.NativeVersion.ShouldBe(new Version(binding.Major, binding.Minor, binding.Build));
	}

	[Fact]
	void Equal_separators_are_a_caller_bug_not_a_verdict() =>
		Should.Throw<ArgumentException>(() =>
			Cast.Int32("42", new NumFormat('.', '.', NumStyles.All)));

	[Fact]
	void Undefined_precision_is_a_caller_bug_not_a_verdict() =>
		Should.Throw<ArgumentOutOfRangeException>(() =>
			Cast.Unix("1700000000", UnixPrecision.Unspecified));
}
